cumulusci/tasks/github/util.py

```diff
@@ -57,123 +60,128 @@ def __call__(
-            "type": git_hash.type,
-        }
-
-    def _create_new_tree_item(self, item):
+    def _create_new_tree_item(self, item: dict) -> Optional[dict]:
         if not item["path"].startswith(self.repo_dir):
             # outside target dir in repo - keep in tree
-            self.logger.debug(
-                "Unchanged (outside target path): {}".format(item["path"])
-            )
-            return item
+            self.logger.debug(f'Unchanged (outside target path): {item["path"]}')
+            return None
 
         local_file, content = self._read_item_content(item)
         new_item = item.copy()
         if content is None:
-            # delete blob from tree
+            # delete blob from tree by setting 'sha' to null
             self.logger.debug(f"Delete: {item['path']}")
-            return content
+            new_item["sha"] = None
         elif self._item_changed(item, content):
-            self.logger.debug("Update: {}".format(local_file))
-            blob_sha = self._create_blob(content, local_file)
-            new_item["sha"] = blob_sha
+            self.logger.debug(f"Update: {local_file}")
+
+            # we need to delete the sha key since GH returns an error when
+            # both 'sha' and 'contents' are passed
+            new_item.pop("sha", None)
+            new_item["content"] = base64.b64encode(content).decode("utf-8")
         else:
-            self.logger.debug("Unchanged: {}".format(item["path"]))
+            self.logger.debug(f'Unchanged: {item["path"]}')
+            return None
         return new_item
 
     def _add_new_files_to_tree(self, new_tree_list):
         new_tree_target_subpaths = [
             self._get_item_sub_path(item)
-            for item in new_tree_list
+            for item in self.tree
             if item["path"].startswith(self.repo_dir)
         ]
 
-        for root, dirs, files in os.walk(self.local_dir):
+        for root, _, files in os.walk(self.local_dir):
             for filename in files:
                 if not filename.startswith("."):
                     local_file = os.path.join(root, filename)
                     local_file_subpath = local_file[(len(self.local_dir) + 1) :]
                     if local_file_subpath not in new_tree_target_subpaths:
                         with io.open(local_file, "rb") as f:
                             content = f.read()
-                        repo_path = (self.repo_dir + "/") if self.repo_dir else ""
+                        repo_path = f"{self.repo_dir}/" if self.repo_dir else ""
                         new_item = {
-                            "path": "{}{}".format(
-                                repo_path, local_file_subpath.replace(os.sep, "/")
-                            ),
-                            "mode": "100644",
+                            "path": f'{repo_path}{local_file_subpath.replace(os.sep, "/")}',
+                            "mode": "100644",  # FIXME: This is wrong
```

Comment:
Why is this wrong? Should we add a WI to correct it?

Reply:
> Why is this wrong?

This assumes that the item (`blob`) is a file, but it could have the executable bit (`100755`) or be a symlink (`120000`), or even a submodule (`160000`). More [here](https://stackoverflow.com/a/69341082).

> Should we add a WI to correct it?

Probably not. This implementation needs these fixes (at least):

1.  The above file mode issue.
2.  It doesn't respect `.gitignore`. We'd need to implement this or add a [library](https://github.com/mherrmann/gitignore_parser/blob/master/gitignore_parser.py).
3.  It ignores any files or directories that start with a dot (`.forceignore`). Simple enough, but we'd need to add an option the consuming task. It'd need to continue ignoring them by default to avoid surprising users accidentally committing code.
4.  It doesn't handle submodules correctly.

None of these are showstoppers for the one task that uses the class, but are probably more impactful downstream in Metecho. Fixing all of these is doable, but I doubt the ROI is worth it, since we can just use Git (via subprocess or a library).

Reply:
Sounds like a design spike is in order for the broad-based fix - could you add that WI?